docs/architecture/front-end-component-structure.md

```diff
@@ -0,0 +1,179 @@
+
+- `ol-common` - For code common to `mit-open`, `ol-page-components` and `ol-components`. Used for TypeScript definitions and anything that has visibility of the app or is not a utility or service.
+
+## Large shared components
+
+Some components are large, contain a lot of functionality or can be considered a “mini app” for embedding within root applications. It may not necessarily be only their size that qualifies them for different treatment - perhaps they have special build chains, are not built on React, or have other requirements. The example in the project is the CKEditor, a rich text editor that has its own workspace, ol-ckeditor, in which we wrap and configure it for use in our application.
+
+These will always have their own package.json (as a defining factor), so are publishable independently to a package repository and should be, for reuse outside the project. To contain these and prevent our top level directories from sprawling, let’s introduce an ol-packages workspace that constitutes our npm packages library. Each of its child directories may contain a workspace of its own.
+
+```text
+├─ ol-packages
+│ ├─ ol-ckeditor – moved from top level
+│ │ ├─ ...
+│ ├─ ol-editable-widget – previously ol-widgets
+│ │ ├─ ...
+│ ├─ ol-rich-text-widget – abstracted from ol-widgets
+│ │ ├─ ...
+│ ├─ ol-embedded-url-widget – abstracted from ol-widgets
+│ │ ├─ ...
+```
+
+## Import rules example
+
+In this sample project:
+
+```text
+├─ project-root
+│ ├─ package.json
+│ ├─ src
+│ │ ├─ App.tsx
+│ │ ├─ pages
+│ │ │ ├─ ArticlePage
+│ │ │ │ ├─ ArticlePage.tsx
+│ │ │ ├─ ErrorPages
+│ │ │ │ ├─ index.ts
+│ │ │ │ ├─ ForbiddenPage.tsx
+│ │ │ │ ├─ NotFoundPage.tsx
+├─ ol-page-components
+│ ├─ package.json
+│ ├─ src
+│ │ ├─ Article
+│ │ │ ├─ Article.tsx
+│ │ │ ├─ Article.test.tsx
+│ │ ├─ CommentBoard
+│ │ │ ├─ CommentBoard.tsx
+│ │ │ ├─ CommentBoard.test.tsx
+│ │ │ ├─ CommentsList.tsx
+├─ ol-components
+│ ├─ package.json
+│ ├─ src
+│ │ ├─ EditableComment
+│ │ │ ├─ EditableComment.tsx
+│ │ │ ├─ EditableComment.test.tsx
+│ │ ├─ SubmitButton
+│ │ │ ├─ SubmitButton.tsx
+│ │ │ ├─ SubmitButton.test.tsx
+├─ ol-services
+│ ├─ package.json
+│ ├─ src
+│ │ ├─ api
+│ │ │ ├─ httpClient.ts
+├─ ol-utilities
+│ ├─ package.json
+│ ├─ src
+│ │ ├─ index.ts
+│ │ ├─ markdownFormatters.ts
+```
+
+- project-root components can import from all `ol-\*` packages.
+
+- `ol-page-components` can import from `ol-components`, `ol-services` and `ol-utilities`, but not project-root.
+
+- `ol-components` can only import from other `ol-components` or `ol-utilities`.
+
+- `ol-services` can only import from `ol-utilities`.
+
+- `ol-utilities` cannot import outside its workspace.
+
+- Directories that contain React components should contain a component file with the same name as the directory that exports a single component.
+
+- Files can only import this component, for example, Article.tsx can import CommentBoard.tsx, but not CommentsList.tsx. We should enforce with lint rules:
```

Comment:
Having looked a bit at ESLint rules for imports, I am not too optimistic that we can actually enforce this. Most rules are based on glob patterns, which aren't flexible enough to enforce that X=Y in `import X from "../whatever/X/Y`.

So, may just have to be unenforced convention.